File: leetcode/LRUCache.java
package leetcode;
class MyDLL <T> {

    static class Node {
        Node prev, next;
        int key;
        int value;

        public Node(int key, int value) {
            this.key = key;
            this.value = value;
        }
    }

    private Node head;
    private Node tail;
    private int size = 0;

    public void print() {
        Node temp = head;
        while (temp != null) {
            System.out.printf("[%d, %d] -> ", temp.key,  temp.value);
            temp = temp.next;
        }
        System.out.println();
    }
    public void add(Node node) {
        //Node node = new Node(key, value);

        if (head == null) {
            head = node;
            tail = node;

        } else {
            tail.next = node;
            node.prev = tail;
            tail = node;
        }
        size++;
    }

    public Node get() {
        return head;
    }


//     public void remove (Node node) {
//         if (node.prev != null) {
//             node.prev.next = node.next;
            
//         }
        
//         if (node.next != null) {
//             node.next.prev = node.prev;
//         }
        
//         if (node.prev == null) {
//             head = node.next;
//         }
        
//         if (node.next == null) {
//             tail = node.prev;
//         }
//         size --;
//     }
    
    public void remove (Node node) {
        // case its empty list 
        
        // -> this cannot happen as Node exists
        
        // just one node
        
        if (head == node && node == tail) {
            head = null;
            tail = null;
            
        }
        
        // case: its head of the list 
        else if ( node == head && node != tail) {
            head = head.next;
            head.prev = null;
            
            // A B <- head
            
        } 
        // case: its tail of the list 
        else if (node == tail && node != head) {
            tail = tail.prev;
            tail.next = null;
        }
        
        // case: its in the middle of the list 
        else {
            node.prev.next = node.next;
            node.next.prev = node.prev;
        }
        
        
        size --;
    }
    

    public int size() {
        return size;
    }
}


public class LRUCache {

    //private List<Pair> queue;
    private MyDLL<MyDLL.Node> queue;

    private Map<Integer, MyDLL.Node> map;
    private int capacity;

    // size 2
    public LRUCache(int capacity) {

        this.capacity = capacity;
        this.map = new HashMap<>();
        //this.queue = new LinkedList<>();
        this.queue = new MyDLL<>();
    }

    public int get(int key) {

        if (map.get(key) == null) {
            // element doesn't exist in the cache
            return -1;
        }
        MyDLL.Node pair = map.get(key);
        queue.remove(pair);
        queue.add(pair);
        return pair.value;
    }

    public void put(int key, int value) {
        if (queue.size() == capacity) {

            if (map.get(key) == null) {
                MyDLL.Node first = queue.get();
                queue.remove(first);
                map.remove(first);
                map.remove(first.key, first);
                System.out.printf("Removed %d with value %d\n" , first.key, first.value);

                print();
                System.out.println("Capacity fixed");
            }
        }

        if (map.get(key) != null) {
            MyDLL.Node removeThis = map.get(key);
            map.remove(key);
            queue.remove(removeThis);
        }

        MyDLL.Node pair = new MyDLL.Node(key, value);
        queue.add(pair);
        map.put(key, pair);
        print();
    }

    private void print() {
        System.out.println("Map");

        for (Integer key: map.keySet()) {
            System.out.printf("map[%d] -> [%d,%d]\n", key, map.get(key).key, map.get(key).value);
        }
        System.out.println();

        System.out.println("Queue");
        queue.print();
        System.out.println();
        System.out.println();
    }
}

File: leetcode/lrucache.py
class LRUCache(object):


    def __init__(self, capacity):
        """
        :type capacity: int
        """
        self.capacity = capacity
        self.dll = DLL()
        self.map = {}
    
    def print_cache(self):
        print("\t Printing cache")
        temp = self.dll.head
        while temp:
            print("\t ", temp.key, temp.value, temp.prev, temp.next)
            temp = temp.get_next()
        print(self.map)
        print("\n\n=====")

    def get(self, key):
        """
        :type key: int
        :rtype: int
        """
        if key in self.map:
            
            self.dll.remove(self.map[key])
            added_node = self.dll.add(self.map[key].key, self.map[key].value)
            self.map[key] = added_node
            return self.map[key].value
        else:
            return -1

    def put(self, key, value):
        """
        :type key: int
        :type value: int
        :rtype: None
        """
        if key in self.map:
            # element already exists in the cache
            # evict only if capacity constraint is reached, which will not happen in this case 
            self.map[key].value = value
            self.get(key)
            return
        
        if len(self.map) == self.capacity:
            removed = self.dll.remove_tail()
            del self.map[removed.key]
            print("Capacity reached, removing item -> ", removed.key, removed.value)
            
        #print("Adding : ", key, value)
        added = self.dll.add(key, value)
        self.map[key] = added
        #print("\tMap: ", self.map)
        
class Node:
        def __init__(self, key, value):
            self.key = key
            self.value = value
            self.next = None
            self.prev = None
        
        def get_prev(self):
            return self.prev

        def set_prev(self, node):
            self.prev = node
        
        def set_next(self, node):
            self.next = node

        def get_next(self):
            return self.next

class DLL(object):
    def __init__(self):
        self.head = None
        self.tail = None

    def remove(self, node):
        if node is None:
            return

        if (node == self.head):
            self.head = node.get_next()
            
            if (self.head):
                self.head.set_prev(None)
            
            
        else:
            node.get_prev().set_next(node.get_next())
            if (node.get_next()):
                node.get_next().set_prev(node.get_prev())
        
        if (node == self.tail):
                self.tail = node.get_prev()

    
    def remove_tail(self):
        response = self.tail
        prev = self.tail.get_prev()
        self.remove(self.tail)
        self.tail = prev

        return response

    
    
    def add(self, key, value):
        if self.head is None:
            self.head = Node(key, value)
            self.tail = self.head
            #print("added 1")
        else:
            #self.head.prev = node
            added = Node(key, value)
            self.head.set_prev(added)
            added.set_next(self.head)
            self.head = added

        return self.head


cache = LRUCache(2)
cache.put(2,1)
cache.print_cache()

cache.put(1,1)
cache.print_cache()

cache.put(2,3)
cache.print_cache()

cache.put(4,1)
cache.print_cache()

print("get", cache.get(1))
cache.print_cache()
print("get", cache.get(2))
cache.print_cache()